fedimint-core/src/query.rs

```diff
@@ -252,6 +256,277 @@ impl<R: Eq + Clone + Debug> QueryStrategy<R> for CurrentConsensus<R> {
+                    .then(|| {
+                        supported_versions
+                            .api
+                            .get_by_major(client_api_version.major)
+                    })
+                    .flatten()
+                    .map(|peer_version| client_api_version.minor <= peer_version.minor)
+                    .unwrap_or(false)
+            })
+            .count();
+
+        if best_major_peer_num < peers_compatible_num {
+            best_major = Some(client_api_version);
+            best_major_peer_num = peers_compatible_num;
+        }
+    }
+
+    best_major
+}
+
+fn discover_common_api_versions_set(
+    client_versions: &SupportedApiVersionsSummary,
+    peer_versions: BTreeMap<PeerId, SupportedApiVersionsSummary>,
+) -> anyhow::Result<ApiVersionSet> {
+    Ok(ApiVersionSet {
+        core: discover_common_core_api_version(
+            &client_versions.core,
+            peer_versions
+                .iter()
+                .map(|(peer_id, peer_supported_api_versions)| {
+                    (*peer_id, peer_supported_api_versions.core.clone())
+                })
+                .collect(),
+        )
+        .ok_or_else(|| format_err!("Could not find a common core API version"))?,
+        modules: client_versions
+            .modules
+            .iter()
+            .filter_map(
+                |(module_instance_id, client_supported_module_api_versions)| {
+                    let discover_common_module_api_version = discover_common_module_api_version(
+                        client_supported_module_api_versions,
+                        peer_versions
+                            .iter()
+                            .filter_map(|(peer_id, peer_supported_api_versions_summary)| {
+                                peer_supported_api_versions_summary
+                                    .modules
+                                    .get(module_instance_id)
+                                    .map(|versions| (*peer_id, versions.clone()))
+                            })
+                            .collect(),
+                    );
+                    discover_common_module_api_version.map(|v| (*module_instance_id, v))
+                },
+            )
+            .collect(),
+    })
+}
+
+impl QueryStrategy<SupportedApiVersionsSummary, ApiVersionSet> for DiscoverApiVersionSet {
+    fn process(
+        &mut self,
+        peer: PeerId,
+        result: api::MemberResult<SupportedApiVersionsSummary>,
+    ) -> QueryStep<ApiVersionSet> {
+        match self.inner.process(peer, result) {
+            QueryStep::Success(o) => {
+                match discover_common_api_versions_set(&self.client_versions, o) {
+                    Ok(o) => QueryStep::Success(o),
+                    // TODO: do we need QueryStep::HighLevelFailureOfSomeKind?
+                    Err(e) => QueryStep::Failure(BTreeMap::from([(
+                        PeerId::from(0),
```

Comment:
```suggestion
                       peer,
```
?

Reply:
The failure here is logical and global, and `0` is used only because we have no error case like that yet, yet we want to put some informative error somewhere.